shell/platform/android/io/flutter/view/VsyncWaiter.java

```diff
@@ -88,19 +102,9 @@ public static void reset() {
   private final FlutterJNI.AsyncWaitForVsyncDelegate asyncWaitForVsyncDelegate =
       new FlutterJNI.AsyncWaitForVsyncDelegate() {
         @Override
-        public void asyncWaitForVsync(long cookie) {
-          Choreographer.getInstance()
-              .postFrameCallback(
-                  new Choreographer.FrameCallback() {
-                    @Override
-                    public void doFrame(long frameTimeNanos) {
-                      long delay = System.nanoTime() - frameTimeNanos;
-                      if (delay < 0) {
-                        delay = 0;
-                      }
-                      flutterJNI.onVsync(delay, refreshPeriodNanos, cookie);
-                    }
-                  });
+        public void asyncWaitForVsync(final long cookie) {
+          frameCallback.cookie = cookie;
```

Comment:
The Java `VsyncWaiter` is a singleton.  But there can be multiple instances of the Flutter engine, each of which has its own `VsyncWaiterAndroid` on the native side.

Each of those `VsyncWaiterAndroid` instances will make JNI calls to `VsyncWaiter.asyncWaitForVsync` and will pass a cookie representing the pointer to the originating `VsyncWaiterAndroid`.

So it is not safe to assume that only one cookie has a pending frame callback.

Reply:
> Is this causing a significant performance issue?
> 
> If the cost of these allocations is noticeable, then one option would be using a collection such as Android's `SparseLongArray` to hold the pending cookies.

You are right. I will collect and dispatch ever single cookie.

Reply:
I have updated commit, please check